Handling webhooks for recurring payments

danielroseman opened the pull request:
* Refactor models to separate Payment from Donation, to allow for multiple payments in a single donation.
* Add support for handling webhooks: payment succeeded, subscription cancelled, payment failed.
* Allow users to change card details on recurring payments, for when payment fails because of an expired card.

Refs #386.

jezdez commented on fundraising/urls.py:
```diff
@@ -9,4 +9,6 @@
     url(r'^thank-you/(?P<donation>[\w]+)/$', views.thank_you, name='thank-you'),
     url(r'^manage-donations/(?P<hero>[\w]+)/$', views.manage_donations, name='manage-donations'),
     url(r'^manage-donations/(?P<hero>[\w]+)/cancel/(?P<donation>[\w]+)$', views.cancel_donation, name='cancel-donation'),
+    url(r'^receive_webhook/', views.receive_webhook, name='receive-webhook'),
```
Mind using dashes here instead of underscores?

danielroseman replied:
Done in 6f7ec7e.

oinopion replied:
`url(r'^receive_webhook/',` :point_right: `url(r'^receive-webhook/',` :wink:

oinopion replied:
BTW don't forget about finishing URL pattern with `$`.

oinopion commented:
Hey @danielroseman, I've somehow missed that you've submitted this PR, sorry for delay. I'll try to review it this evening. Thanks for handling this :raised_hands:

danielroseman commented:
@oinopion no problem, I didn't finish it until after you'd left on Friday.

I realized that we needed some way of updating the card details if the payment failed, so I've implemented that with a button from the manage donations page (the button is currently unstyled). Let me know if you'd prefer that change split out into a separate PR.

danielroseman commented:
@oinopion Will you have a chance to look at this?

oinopion commented on djangoproject/static/js/mod/stripe-change-card.js:
```diff
@@ -0,0 +1,43 @@
+    'stripe-checkout'
+], function ($) {
+    var $heroForm = $('.django-hero-form')
+    $('.django-hero-form').on('click', '.change-card', function() {
```
You can use $heroForm instead of selecting it again:

    $heroForm.on('click' ...

danielroseman replied:
Done.

oinopion commented on djangoproject/static/js/mod/stripe-change-card.js:
```diff
@@ -0,0 +1,43 @@
+    var $heroForm = $('.django-hero-form')
+    $('.django-hero-form').on('click', '.change-card', function() {
+        $this = $(this);
+        var donation_id = $this.data('donation_id');
```
In JS we use camelCase, so `var donation_id` :arrow_right: `var donationId`.

The `data` accessor should also use camel case, so `$this.data('donationId')`. 
(This accessor is special, because in HTML you specify it with dashes, see other comment).

danielroseman replied:
Done.

oinopion commented on djangoproject/templates/fundraising/manage-donations.html:
```diff
@@ -24,6 +25,9 @@
       <h3>{{ form.instance.campaign }}</h3>
       {% include 'fundraising/includes/_form.html' %}
+      <div class="change-card-container">
+        <input type="button" value="Change card details" class="change-card" data-donation_id="{{ form.instance.id }}" data-donor-email="{{ hero.email }}"> <span class="change-card-result"></span>
```
Change `data-donation_id` to `data-donation-id` and in relevant JS `donation_id` to `donationId`. :sunny:

danielroseman replied:
Done.

oinopion commented on fundraising/forms.py:
```diff
@@ -253,18 +253,24 @@ def make_donation(self):
+                donation_params['subscription_amount'] = amount
             donation = Donation.objects.create(**donation_params)
 
+            Payment.objects.create(**{
```
There's no need to use dict unpacking. Could you change to just normal keywords arguments?

danielroseman replied:
Done.

oinopion commented on fundraising/migrations/0032_auto_20150604_0813.py:
```diff
@@ -0,0 +1,29 @@
+            old_name='amount',
+            new_name='subscription_amount',
+        ),
+        migrations.AlterField(
```
Shouldn't we set NULL to all rows where is `onetime`?

danielroseman replied:
Done in new migration 0033.

oinopion commented on fundraising/models.py:
```diff
@@ -109,22 +109,31 @@ def __str__(self):
 
 
+class Payment(models.Model):
+    donation = models.ForeignKey(Donation)
```
Can you add explicit `related_name="payments"`?

danielroseman replied:
Not sure why this is necessary. We don't do it for any other reverse relations in this app.

oinopion commented on fundraising/tests.py:
```diff
@@ -109,7 +111,8 @@ def test_submitting_donation_form(self, charge_create, customer_create):
         self.assertEqual(donations.count(), 1)
-        self.assertEqual(donations[0].amount, 100)
+        self.assertEqual(donations[0].subscription_amount, None)
+        self.assertEqual(donations[0].payment_set.first().amount, 100)
```
Looks like it would be good to have `donation.total_payments()` method that returns sum of all payments. Would make tests here more clear.

danielroseman replied:
Done.

oinopion commented on fundraising/tests.py:
```diff
@@ -382,3 +379,40 @@ def test_create_hero_for_donation_with_campaign(self):
+        )
+
+    def post_data(self, filename):
+        with open('fundraising/test_data/{}.json'.format(filename)) as f:
```
Would be best to specify filename by absolute path, based either on `settings.BASE_DIR` or current `__file__`:

    file_path = os.path.join(settings.BASE_DIR, 'fundraising', 'test_data', '{}.json'.format(filename))
    with open(file_path):
        ...

danielroseman replied:
Actually settings.BASE_DIR is a PosixPath object, so I've used the methods of that class.

oinopion commented on fundraising/tests.py:
```diff
@@ -382,3 +379,40 @@ def test_create_hero_for_donation_with_campaign(self):
+
+    def post_data(self, filename):
+        with open('fundraising/test_data/{}.json'.format(filename)) as f:
+            response = self.client.post(
```
You can return straight from within `with` block, no need for new variable:
    
    with open(...):
        return self.client.post(...)

danielroseman replied:
Done.

oinopion commented on fundraising/views.py:
```diff
@@ -86,6 +91,8 @@ def manage_donations(request, hero):
     hero = get_object_or_404(DjangoHero, pk=hero)
     recurring_donations = hero.donation_set.filter(
```
Would be really nice if we contained this query in a method on `DjangoHero` class.

danielroseman replied:
Since this query will be simplified when we remove the null=True after Tim's comment, this doesn't seem to be worth doing.

oinopion commented on fundraising/views.py:
```diff
@@ -131,3 +154,64 @@ def cancel_donation(request, hero, donation):
+            'invoice.payment_failed': self.payment_failed,
+            'customer.subscription.deleted': self.subscription_cancelled,
+        }
+        handler = handlers.get(self.event.type, lambda: HttpResponse(404))
```
I think it would be more correct to return 422 -- unprocessable entity.

danielroseman replied:
Done.

oinopion commented on fundraising/views.py:
```diff
@@ -131,3 +154,64 @@ def cancel_donation(request, hero, donation):
+@csrf_exempt
+def receive_webhook(request):
+    data = json.loads(request.body.decode())
+    event = stripe.resource.convert_to_stripe_object(data, stripe.api_key)
```
We need to figure out a way to secure this: anyone can check source code and send something to this view.
I see two possible solutions:
1. Have this URL randomized with some token saved in settings/env. variable, set this url in stripe:
    
    # urls.py
        url(r'^receive-webhook/(?P<token>[\w]+)/$', views.receive_webhook, name='receive-webhook'),
    # views.py
    def receive_webhook(request, token):
         if token != settings.STRIPE_SECRET_WEBHOOK_TOKEN:
             return HttpNotFound()
       ...

2. Use API instead of trusting submitted data, as suggested here: https://stripe.com/docs/webhooks#receiving-a-webhook

danielroseman replied:
Really great point, thanks. I've gone with the second option - we now explicitly re-request any event that we receive as a webhook. View and tests updated.

oinopion commented on fundraising/views.py:
```diff
@@ -131,3 +154,64 @@ def cancel_donation(request, hero, donation):
+        subscription = self.event.data.object
+        donation = get_object_or_404(
+            Donation, stripe_subscription_id=subscription.id)
+        donation.stripe_subscription_id = ''
```
Would be cleaner to set it to `None`. And if we're at that, let's make sure this field is `blank=True, null=True` :smiley_cat:

timgraham replied:
I think it's bad to have `null=True` on `CharField` -- there are two "empty values" and any "nulls" get converted to empty string when saving in the admin.

oinopion replied:
So you're suggesting to convert it to `null=False, blank=True, default=''. This probably needs a migration.

timgraham replied:
Correct -- this could be done separately to keep the fundraising-recurrsing branch from containing so many changes.

danielroseman replied:
I'll do this for all fundraising CharFields in a subsequent PR.

oinopion commented:
@danielroseman excellent work with this PR :ok_hand: 
I've left some comments, mostly stylistic.
This one is important: https://github.com/django/djangoproject.com/pull/486#discussion_r32177238

Thanks for doing that, I can't stress enough how invaluable your stripe expertise is to us! :sunny: :+1: :raised_hands:

danielroseman commented:
Thanks for the great review @oinopion - I've addressed all the comments and in particular added some code to re-request the events from Stripe for security.

I'll follow up with another PR to fix the CharField null=True issue when this is merged.

olasitarska commented:
@oinopion @timgraham any objections to merging it?;) :shipit:

timgraham commented on djangoproject/static/js/mod/stripe-change-card.js:
```diff
@@ -0,0 +1,43 @@
+define([
+    'jquery', //requires jquery
+    'stripe-checkout'
+], function ($) {
```
I think we should be consistent with no space after function?

danielroseman replied:
Fixed.

timgraham commented on djangoproject/static/js/mod/stripe-change-card.js:
```diff
@@ -0,0 +1,43 @@
+    'jquery', //requires jquery
+    'stripe-checkout'
+], function ($) {
+    var $heroForm = $('.django-hero-form')
```
missing ;

danielroseman replied:
Fixed.

timgraham commented on djangoproject/templates/fundraising/email/payment_failed.txt:
```diff
@@ -0,0 +1,13 @@
+Thanks very much for your support.
+
+Regards,
+The Django Project
```
"Django Software Foundation" as below? + chop blank line

danielroseman replied:
Fixed.

timgraham commented on djangoproject/templates/fundraising/email/payment_failed.txt:
```diff
@@ -0,0 +1,13 @@
+Software Foundation has failed. This might be because your payment card has 
+expired.
+
+Please go to {{ current_site }}{% url 'fundraising:manage-donations' donation.donor.pk %} to add a new card. We won't be able to take any payments in the mean time.
```
what's `current_site`? I think {% url %} includes the scheme and domain because we're using django-hosts (at least, I saw this behavior in the password reset email. is this template tested?)

danielroseman replied:
Good point, I've removed the `current_site`. I've added some checks in the tests to make sure the full URL is being used.

timgraham commented on djangoproject/templates/fundraising/manage-donations.html:
```diff
@@ -6,7 +7,7 @@
   Thank you!</p>
 
-<form enctype="multipart/form-data" action="" method="post" class="form-input django-hero-form">
+<form enctype="multipart/form-data" action="" method="post" class="form-input django-hero-form" data-stripe-key="{{ stripe_publishable_key }}" data-stripe-icon="{% static 'img/dj-stripe-icon.jpg' %}" data-update-card-url="{% url 'fundraising:update-card' %}">
```
could you multiline this so it fits in 119 chars so review doesn't require horizontal scrolling in GitHub

danielroseman replied:
Done.

timgraham commented on djangoproject/templates/fundraising/manage-donations.html:
```diff
@@ -24,6 +25,9 @@
       <h3>{{ form.instance.campaign }}</h3>
       {% include 'fundraising/includes/_form.html' %}
+      <div class="change-card-container">
+        <input type="button" value="Change card details" class="change-card" data-donation-id="{{ form.instance.id }}" data-donor-email="{{ hero.email }}"> <span class="change-card-result"></span>
```
too long to review easily

danielroseman replied:
Done.

timgraham commented on fundraising/tests.py:
```diff
@@ -382,3 +379,60 @@ def test_create_hero_for_donation_with_campaign(self):
+        return self.client.post(
+            reverse('fundraising:receive-webhook'),
+            data='{"id": "evt_12345"}',
+            content_type='application/json'
```
trailing ,

danielroseman replied:
Fixed.

timgraham commented on fundraising/views.py:
```diff
@@ -131,3 +155,69 @@ def cancel_donation(request, hero, donation):
+@require_POST
+@csrf_exempt
+def receive_webhook(request):
+    data = json.loads(request.body.decode())
```
might be worth being a bit defensive in json parsing so silly people don't trigger 500s and fill our error logs.

danielroseman replied:
Done.

danielroseman commented:
@timgraham @oinopion @olasitarska can this be merged?

oinopion commented:
@danielroseman merged :boom: :bow: :sparkles: 

Thanks for hard work and sorry for delay -- I never have as much time for open source as I'd like to.
